FirebaseMLModelDownloader/Sources/ModelDownloadManager.swift

```diff
@@ -0,0 +1,143 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseCore
+
+enum DownloadStatus {
+  case notStarted
+  case inProgress
+  case completed
+}
+
+/// Progress and completion handlers for a model download.
+class DownloadHandlers: NSObject {
+  typealias ProgressHandler = (Float) -> Void
+  typealias Completion = (Result<CustomModel, DownloadError>) -> Void
+
+  var progressHandler: ProgressHandler?
+  var completion: Completion
+
+  init(progressHandler: ProgressHandler?, completion: @escaping Completion) {
+    self.progressHandler = progressHandler
+    self.completion = completion
+  }
+}
+
+/// Manager for model downloads.
+class ModelDownloadManager: NSObject {
+  let app: FirebaseApp
+  var modelInfo: ModelInfo
+  var taskHandlers: [URLSessionDownloadTask: DownloadHandlers] = [:]
+
+  private(set) var downloadStatus: DownloadStatus = .notStarted
+
+  private lazy var downloadSession = URLSession(configuration: .ephemeral,
+                                                delegate: self,
+                                                delegateQueue: nil)
+
+  init(app: FirebaseApp, modelInfo: ModelInfo) {
+    self.app = app
+    self.modelInfo = modelInfo
+  }
+
+  /// Associate progress and completion handlers with download task.
+  private func setHandlers(for downloadTask: URLSessionDownloadTask, handlers: DownloadHandlers) {
+    taskHandlers[downloadTask] = handlers
+  }
+
+  /// Get download handlers for the given download task.
+  private func getHandlers(for downloadTask: URLSessionDownloadTask) -> DownloadHandlers? {
+    return taskHandlers[downloadTask]
+  }
+
+  /// Asynchronously download model file to device.
+  func startModelDownload(url: URL, progressHandler: DownloadHandlers.ProgressHandler? = nil,
+                          completion: @escaping DownloadHandlers.Completion) {
+    let downloadTask = downloadSession.downloadTask(with: url)
+    let downloadHandlers = DownloadHandlers(
+      progressHandler: progressHandler,
+      completion: completion
+    )
+    setHandlers(for: downloadTask, handlers: downloadHandlers)
+    downloadTask.resume()
+    downloadStatus = .inProgress
+  }
+}
+
+/// Extension to handle delegate methods.
+extension ModelDownloadManager: URLSessionDownloadDelegate {
+  func urlSession(_ session: URLSession,
+                  downloadTask: URLSessionDownloadTask,
+                  didFinishDownloadingTo location: URL) {
+    guard let handlers = getHandlers(for: downloadTask) else { return }
+    downloadStatus = .completed
+    let savedURL = ModelFileManager.modelsDirectory
+      .appendingPathComponent(downloadedModelFileName)
+    do {
+      try ModelFileManager.moveFile(at: location, to: savedURL)
+    } catch {
+      handlers.completion(.failure(.internalError(description: error.localizedDescription)))
+      return
+    }
+
+    modelInfo.path = savedURL.absoluteString
```

Comment:
For the model file to be downloaded, I need the model download URL to already be stored in the modelInfo object (along with name, size, and hash). The `path` property (which is the local path of the model file on device) is the only one that should be set when the actual model file is downloaded. Ideally these are sequential operations, and all these properties are backed by UserDefaults. However, let's chat through any concerns and I'm open to suggestions to changing this!

Reply:
My concern is that I cannot see(maybe I'm just missing) anything in the code that prevents the model info being concurrently updated (e.g. several downloads may be performed, model info may be modified at a different place, etc.). It means a developer modifying the code must be aware of the intentions you've describe in above. Practically it means the the developer would have to examine all places where the model info is accessed when modifying any related logic, which opens additional room for errors. My suggestion would be to put a bit extra effort in designing the API so that the "illegal" operations are not possible to perform, which usually saves significant time in the future cutting down maintenance cost. 

Specifically to this case - the current API allows setting any property of `ModelInfo` instance at any moment even during download. Also the UserDefaults allows to change the property value of `ModelInfo` object without accessing the instance. This opens room for a lot of corner cases, inconsistent states and unexpected behaviour that may be introduced accidentally.